pkg/querier/queryrange/prometheus.go

```diff
@@ -41,41 +43,91 @@ func (PrometheusExtractor) ResponseWithoutHeaders(resp queryrange.Response) quer
 // encode encodes a Prometheus response and injects Loki stats.
 func (p *LokiPromResponse) encode(ctx context.Context) (*http.Response, error) {
 	sp := opentracing.SpanFromContext(ctx)
+	var (
+		b   []byte
+		err error
+	)
+	if p.Response.Data.ResultType == loghttp.ResultTypeVector {
+		b, err = p.marshalVector()
+	} else {
+		b, err = p.marshalMatrix()
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if sp != nil {
+		sp.LogFields(otlog.Int("bytes", len(b)))
+	}
+
+	resp := http.Response{
+		Header: http.Header{
+			"Content-Type": []string{"application/json"},
+		},
+		Body:       ioutil.NopCloser(bytes.NewBuffer(b)),
+		StatusCode: http.StatusOK,
+	}
+	return &resp, nil
+}
+
+func (p *LokiPromResponse) marshalVector() ([]byte, error) {
+	vec := make(loghttp.Vector, len(p.Response.Data.Result))
+	for i, v := range p.Response.Data.Result {
+		lbs := make(model.LabelSet, len(v.Labels))
+		for _, v := range v.Labels {
+			lbs[model.LabelName(v.Name)] = model.LabelValue(v.Value)
+		}
+		vec[i] = model.Sample{
+			Metric:    model.Metric(lbs),
+			Timestamp: model.Time(v.Samples[0].TimestampMs),
+			Value:     model.SampleValue(v.Samples[0].Value),
+		}
+	}
+	return jsonStd.Marshal(struct {
```

Comment:
I think we should define this struct as a type somewhere and use it here?

Reply:
sounds good will do later.